Use restricted Damerau-Levenshtein algorithm

jhpratt opened the pull request:
This uses the same implementation as the one used in rustc, so review should be simple. As with rust-lang/rust#108200, the module and function names have been changed to be implementation-agnostic.

[Reference](https://github.com/rust-lang/rust/blob/13d1802b8882452f7d9d1bf514a096c5c8a22303/compiler/rustc_span/src/edit_distance.rs) for rustc's current implementation.

epage commented on src/cargo/util/lev_distance.rs:
```diff
@@ -1,93 +0,0 @@
-use std::cmp;
-
-pub fn lev_distance(me: &str, t: &str) -> usize {
```
In the future, I'd recommend doing the rename and algorithm changes in separate commits so it is easier to review

epage commented on src/cargo/core/workspace.rs:
```diff
@@ -1245,8 +1245,9 @@ impl<'cfg> Workspace<'cfg> {
 
-        let levenshtein_test =
-            |a: InternedString, b: InternedString| lev_distance(a.as_str(), b.as_str()) < 4;
+        let levenshtein_test = |a: InternedString, b: InternedString| {
```
Since you were updating references to the algorithm, should this one also be updated?

jhpratt commented:
I've split the commit into two and changed the remaining instances of the specific name (the one you caught and one I missed).

Assuming CI passes again, this should be good to go.

bors commented:
:pushpin: Commit ccd77a3efc6c608a0c8a677204e541eb4163285b has been approved by `epage`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=epage ccd77a3efc6c608a0c8a677204e541eb4163285b -->
<!-- homu: {"type":"Approved","sha":"ccd77a3efc6c608a0c8a677204e541eb4163285b","approver":"epage","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit ccd77a3efc6c608a0c8a677204e541eb4163285b with merge d2e7cfb2487e93a41e8ad93359d245420cce86c0...
<!-- homu: {"type":"BuildStarted","head_sha":"ccd77a3efc6c608a0c8a677204e541eb4163285b","merge_sha":"d2e7cfb2487e93a41e8ad93359d245420cce86c0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/4700897442/jobs/8337229165)
Approved by: epage
Pushing d2e7cfb2487e93a41e8ad93359d245420cce86c0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"epage","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/4700897442/jobs/8337229165"},"merge_sha":"d2e7cfb2487e93a41e8ad93359d245420cce86c0"} -->
